File: Bookcase web/src/test/java/cz/vhromada/book/web/common/BookUtils.java
package cz.vhromada.book.web.common;

import static org.assertj.core.api.SoftAssertions.assertSoftly;

import cz.vhromada.book.entity.Book;
import cz.vhromada.book.web.fo.BookFO;
import cz.vhromada.common.Language;
import cz.vhromada.common.utils.CollectionUtils;

/**
 * A class represents utility class for books.
 *
 * @author Vladimir Hromada
 */
public final class BookUtils {

    /**
     * Creates a new instance of BookUtils.
     */
    private BookUtils() {
    }

    /**
     * Returns FO for book.
     *
     * @return FO for book
     */
    public static BookFO getBookFO() {
        final BookFO book = new BookFO();
        book.setId(1);
        book.setCzechName("Czech name");
        book.setOriginalName("Original name");
        book.setLanguages(CollectionUtils.newList(Language.CZ));
        book.setIsbn("ISBN");
        book.setIssueYear("2000");
        book.setDescription("Description");
        book.setElectronic(true);
        book.setPaper(false);
        book.setNote("Note");
        book.setPosition(0);
        book.setAuthors(CollectionUtils.newList(1));
        book.setCategories(CollectionUtils.newList(1));

        return book;
    }

    /**
     * Returns book.
     *
     * @return book
     */
    public static Book getBook() {
        final int issueYear = 2000;

        final Book book = new Book();
        book.setId(1);
        book.setCzechName("Czech name");
        book.setOriginalName("Original name");
        book.setLanguages(CollectionUtils.newList(Language.CZ));
        book.setIsbn("ISBN");
        book.setIssueYear(issueYear);
        book.setDescription("Description");
        book.setElectronic(true);
        book.setPaper(false);
        book.setNote("Note");
        book.setPosition(0);
        book.setAuthors(CollectionUtils.newList(AuthorUtils.getAuthor()));
        book.setCategories(CollectionUtils.newList(CategoryUtils.getCategory()));

        return book;
    }

    /**
     * Asserts book deep equals.
     *
     * @param expected expected FO for book
     * @param actual   actual book
     */
    @SuppressWarnings("Duplicates")
    public static void assertBookDeepEquals(final BookFO expected, final Book actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(actual.getId()).isEqualTo(expected.getId());
            softly.assertThat(actual.getCzechName()).isEqualTo(expected.getCzechName());
            softly.assertThat(actual.getOriginalName()).isEqualTo(expected.getOriginalName());
            softly.assertThat(actual.getLanguages())
                .hasSameSizeAs(expected.getLanguages())
                .hasSameElementsAs(expected.getLanguages());
            softly.assertThat(actual.getIsbn()).isEqualTo(expected.getIsbn());
            softly.assertThat(actual.getIssueYear()).isEqualTo(Integer.parseInt(expected.getIssueYear()));
            softly.assertThat(actual.getDescription()).isEqualTo(expected.getDescription());
            softly.assertThat(actual.getElectronic()).isEqualTo(expected.getElectronic());
            softly.assertThat(actual.getPaper()).isEqualTo(expected.getPaper());
            softly.assertThat(actual.getNote()).isEqualTo(expected.getNote());
            softly.assertThat(actual.getPosition()).isEqualTo(expected.getPosition());
            AuthorUtils.assertAuthorsDeepEquals(expected.getAuthors(), actual.getAuthors());
            CategoryUtils.assertCategoriesDeepEquals(expected.getCategories(), actual.getCategories());
        });
    }

    /**
     * Asserts book deep equals.
     *
     * @param expected expected book
     * @param actual   actual FO for book
     */
    @SuppressWarnings("Duplicates")
    public static void assertBookDeepEquals(final Book expected, final BookFO actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(actual.getId()).isEqualTo(expected.getId());
            softly.assertThat(actual.getCzechName()).isEqualTo(expected.getCzechName());
            softly.assertThat(actual.getOriginalName()).isEqualTo(expected.getOriginalName());
            softly.assertThat(actual.getLanguages())
                .hasSameSizeAs(expected.getLanguages())
                .hasSameElementsAs(expected.getLanguages());
            softly.assertThat(actual.getIsbn()).isEqualTo(expected.getIsbn());
            softly.assertThat(actual.getIssueYear()).isEqualTo(Integer.toString(expected.getIssueYear()));
            softly.assertThat(actual.getDescription()).isEqualTo(expected.getDescription());
            softly.assertThat(actual.getElectronic()).isEqualTo(expected.getElectronic());
            softly.assertThat(actual.getPaper()).isEqualTo(expected.getPaper());
            softly.assertThat(actual.getNote()).isEqualTo(expected.getNote());
            softly.assertThat(actual.getPosition()).isEqualTo(expected.getPosition());
            AuthorUtils.assertAuthorListDeepEquals(expected.getAuthors(), actual.getAuthors());
            CategoryUtils.assertCategoryListDeepEquals(expected.getCategories(), actual.getCategories());
        });
    }

}

File: bookcase-web/src/test/kotlin/cz/vhromada/bookcase/web/common/BookUtils.kt
package cz.vhromada.bookcase.web.common

import cz.vhromada.bookcase.entity.Book
import cz.vhromada.bookcase.web.fo.BookFO
import org.assertj.core.api.SoftAssertions.assertSoftly

/**
 * A class represents utility class for books.
 *
 * @author Vladimir Hromada
 */
object BookUtils {

    /**
     * Returns FO for book.
     *
     * @return FO for book
     */
    fun getBookFO(): BookFO {
        return BookFO(
                id = 1,
                czechName = "Czech name",
                originalName = "Original name",
                isbn = "ISBN",
                issueYear = "2000",
                description = "Description",
                note = "Note",
                position = 0,
                authors = listOf(1),
                categories = listOf(1))
    }

    /**
     * Returns book.
     *
     * @return book
     */
    fun getBook(): Book {
        return Book(
                id = 1,
                czechName = "Czech name",
                originalName = "Original name",
                isbn = "ISBN",
                issueYear = 2000,
                description = "Description",
                note = "Note",
                position = 0,
                authors = listOf(AuthorUtils.getAuthor()),
                categories = listOf(CategoryUtils.getCategory()))
    }

    /**
     * Asserts book deep equals.
     *
     * @param expected expected FO for book
     * @param actual   actual book
     */
    fun assertBookDeepEquals(expected: BookFO?, actual: Book?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertSoftly {
            it.assertThat(actual!!.id).isEqualTo(expected!!.id)
            it.assertThat(actual.czechName).isEqualTo(expected.czechName)
            it.assertThat(actual.originalName).isEqualTo(expected.originalName)
            it.assertThat(actual.isbn).isEqualTo(expected.isbn)
            it.assertThat(actual.issueYear).isEqualTo(expected.issueYear!!.toInt())
            it.assertThat(actual.description).isEqualTo(expected.description)
            it.assertThat(actual.note).isEqualTo(expected.note)
            it.assertThat(actual.position).isEqualTo(expected.position)
            it.assertThat(actual.authors).isNull()
            it.assertThat(actual.categories).isNull()
        }
    }

    /**
     * Asserts book deep equals.
     *
     * @param expected expected FO for book
     * @param actual   actual book
     */
    fun assertBookDeepEquals(expected: Book?, actual: BookFO?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertSoftly {
            it.assertThat(actual!!.id).isEqualTo(expected!!.id)
            it.assertThat(actual.czechName).isEqualTo(expected.czechName)
            it.assertThat(actual.originalName).isEqualTo(expected.originalName)
            it.assertThat(actual.isbn).isEqualTo(expected.isbn)
            it.assertThat(actual.issueYear).isEqualTo(expected.issueYear.toString())
            it.assertThat(actual.description).isEqualTo(expected.description)
            it.assertThat(actual.note).isEqualTo(expected.note)
            it.assertThat(actual.position).isEqualTo(expected.position)
            AuthorUtils.assertAuthorsDeepEquals(expected.authors, actual.authors)
            CategoryUtils.assertCategoriesDeepEquals(expected.categories, actual.categories)
        }
    }

}
